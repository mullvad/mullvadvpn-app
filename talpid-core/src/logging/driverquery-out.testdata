"Module Name","Display Name","Description","Driver Type","Start Mode","State","Status","Accept Stop","Accept Pause","Paged Pool(bytes)","Code(bytes)","BSS(bytes)","Link Date","Path","Init(bytes)"
"ACPI","Microsoft ACPI Driver","Microsoft ACPI Driver","Kernel ","Boot","Running","OK","TRUE","FALSE","184 320","466 944","0","","C:\WINDOWS\system32\drivers\ACPI.sys","24 576"
"mullvad-split-tunnel","Mullvad Split Tunnel Service","Mullvad Split Tunnel Service","Kernel ","Manual","Stopped","OK","FALSE","FALSE","4 096","53 248","0","2022-09-22 11:41:33","\??\C:\Program Files\Mullvad VPN\resources\mullvad-split-tunnel.sys","8 192"
"MullvadWireGuard","MullvadWireGuard","MullvadWireGuard","Kernel ","Manual","Stopped","OK","FALSE","FALSE","0","131 072","0","2024-06-04 14:21:55","C:\WINDOWS\system32\drivers\mullvad-wireguard.sys","12 288"
"NDIS","NDIS System Driver","NDIS System Driver","Kernel ","Boot","Running","OK","TRUE","FALSE","331 776","1 028 096","0","","C:\WINDOWS\system32\drivers\ndis.sys","16 384"
