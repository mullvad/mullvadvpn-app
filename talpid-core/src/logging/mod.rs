use std::{fs, io, path::Path};

/// Unable to create new log file
#[derive(err_derive::Error, Debug)]
#[error(display = "Unable to create new log file")]
pub struct RotateLogError(#[error(source)] io::Error);

/// Create a new log file while backing up a previous version of it.
///
/// A new log file is created with the given file name, but if a file with that name already exists
/// it is backed up with the extension changed to `.old.log`.
pub fn rotate_log(file: &Path) -> Result<(), RotateLogError> {
    let backup = file.with_extension("old.log");
    if let Err(error) = fs::rename(file, &backup) {
        if error.kind() != io::ErrorKind::NotFound {
            log::warn!(
                "Failed to rotate log file to {}: {}",
                backup.display(),
                error
            );
        }
    }

    fs::File::create(file).map(|_| ()).map_err(RotateLogError)
}
