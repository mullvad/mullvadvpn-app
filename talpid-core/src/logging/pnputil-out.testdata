InstanceId,DeviceDescription,ClassName,ClassGuid,ManufacturerName,Status,ProblemCode,ProblemStatus,DriverName,ExtensionDriverNames
"SWD\MSRRAS\MS_PPPOEMINIPORT","WAN Miniport (PPPOE)","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","netrasa.inf",""
"SWD\MSRRAS\MS_PPTPMINIPORT","WAN Miniport (PPTP)","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","netrasa.inf",""
"SWD\MSRRAS\MS_AGILEVPNMINIPORT","WAN Miniport (IKEv2)","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","netavpna.inf",""
"ROOT\KDNIC\0000","Microsoft Kernel Debug Network Adapter","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","kdnic.inf",""
"USB\VID_0BDA&PID_8153\001000001","Realtek USB GbE Family Controller","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Realtek","Disconnected","","","oem1.inf",""
"BTH\MS_BTHPAN\8&19b34d4&0&2","Bluetooth Device (Personal Area Network)","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","bthpan.inf",""
"ROOT\VMS_MP\0000","Hyper-V Virtual Ethernet Adapter #3","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","wvms_mp_windows.inf",""
"ROOT\WINTUN\0000","Wintun Userspace Tunnel","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","WireGuard LLC","Started","","","oem2.inf",""
"SWD\MSRRAS\MS_NDISWANIPV6","WAN Miniport (IPv6)","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","netrasa.inf",""
"SWD\MSRRAS\MS_L2TPMINIPORT","WAN Miniport (L2TP)","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","netrasa.inf",""
"ROOT\VMS_VSMP\0000","Hyper-V Virtual Switch Extension Adapter","Net","{4d36e972-e325-11ce-bfc1-08002be10318}","Microsoft","Started","","","wvms_mp_windows.inf",""
