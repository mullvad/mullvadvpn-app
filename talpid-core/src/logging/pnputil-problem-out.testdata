InstanceId,DeviceDescription,ClassName,ClassGuid,ManufacturerName,Status,ProblemCode,ProblemStatus,DriverName,ExtensionDriverNames
"UEFI\RES_{redacted}\0","Camera Firmware","Firmware","{f2e7dd72-6468-4e36-b6f1-6488f42c1b52}","Realtek Semiconductor Corp.","Problem","10","0xC0210000","oem84.inf",""
