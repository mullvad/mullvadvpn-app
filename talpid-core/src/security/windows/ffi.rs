use libc::{c_char, c_void};

pub type ErrorSink = extern "system" fn(msg: *const c_char, ctx: *mut c_void);

pub extern "system" fn error_sink(msg: *const c_char, _ctx: *mut c_void) {
    use std::ffi::CStr;
    if msg.is_null() {
        error!("Log message from FFI boundary is NULL");
    } else {
        error!("{}", unsafe { CStr::from_ptr(msg).to_string_lossy() });
    }
}

#[macro_export]
macro_rules! ffi_error {
    ($result:ident, $error:expr) => {
        #[repr(C)]
        #[derive(Debug)]
        pub struct $result {
            success: bool,
        }

        impl $result {
            pub fn into_result(self) -> Result<()> {
                match self.success {
                    true => Ok(()),
                    false => Err($error),
                }
            }
        }

        impl Into<Result<()>> for $result {
            fn into(self) -> Result<()> {
                self.into_result()
            }
        }
    };
}
