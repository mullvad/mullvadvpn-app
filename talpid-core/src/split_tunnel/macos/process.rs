//! This module keeps tracks of maintains a list of processes, and keeps it up to date by observing
//! the syscalls `fork`, `exec`, and `exit`.
//! Each process has an exclusion state, based on which paths the process monitor is instructed to
//! exclude.
//! The module currently relies on the `eslogger` tool to do so, which in turn relies on the
//! Endpoint Security framework.

use futures::channel::oneshot;
use libc::{proc_listallpids, proc_pidpath};
use serde::Deserialize;
use std::{
    collections::{HashMap, HashSet},
    ffi::c_void,
    io,
    path::PathBuf,
    process::Stdio,
    ptr,
    sync::{Arc, LazyLock, Mutex},
    time::Duration,
};
use talpid_platform_metadata::MacosVersion;
use talpid_types::tunnel::ErrorStateCause;
use tokio::io::{AsyncBufReadExt, BufReader};

const SHUTDOWN_TIMEOUT: Duration = Duration::from_secs(3);
const EARLY_FAIL_TIMEOUT: Duration = Duration::from_millis(500);

static MIN_OS_VERSION: LazyLock<MacosVersion> =
    LazyLock::new(|| MacosVersion::from_raw_version("13.0.0").unwrap());

#[derive(thiserror::Error, Debug)]
pub enum Error {
    /// Only macOS 13 and later is supported
    #[error("Unsupported macOS version: {actual}, expected at least {}", *MIN_OS_VERSION)]
    UnsupportedMacosVersion {
        actual: talpid_platform_metadata::MacosVersion,
    },
    /// Failed to start eslogger listener
    #[error("Failed to start eslogger")]
    StartMonitor(#[source] io::Error),
    /// The app requires TCC approval from the user.
    #[error("The app needs TCC approval from the user for Full Disk Access")]
    NeedFullDiskPermissions,
    /// eslogger failed due to an unknown error
    #[error("eslogger returned an error")]
    MonitorFailed(#[source] io::Error),
    /// Monitor task panicked
    #[error("Monitor task panicked")]
    MonitorTaskPanicked(#[source] tokio::task::JoinError),
    /// Failed to list processes
    #[error("Failed to list processes")]
    InitializePids(#[source] io::Error),
    /// Failed to find path for a process
    #[error("Failed to find path for a process: {}", _0)]
    FindProcessPath(#[source] io::Error, u32),
}

impl From<&Error> for ErrorStateCause {
    fn from(value: &Error) -> Self {
        match value {
            Error::NeedFullDiskPermissions => ErrorStateCause::NeedFullDiskPermissions,
            _ => ErrorStateCause::SplitTunnelError,
        }
    }
}

pub struct ProcessMonitor(());

#[derive(Debug)]
pub struct ProcessMonitorHandle {
    stop_proc_tx: Option<oneshot::Sender<oneshot::Sender<()>>>,
    proc_task: tokio::task::JoinHandle<Result<(), Error>>,
    states: ProcessStates,
}

impl ProcessMonitor {
    pub async fn spawn() -> Result<ProcessMonitorHandle, Error> {
        check_os_version_support()?;
        let states = ProcessStates::new()?;

        let proc = spawn_eslogger()?;
        let (stop_proc_tx, stop_rx): (_, oneshot::Receiver<oneshot::Sender<_>>) =
            oneshot::channel();
        let mut proc_task = tokio::spawn(handle_eslogger_output(proc, states.clone(), stop_rx));

        match tokio::time::timeout(EARLY_FAIL_TIMEOUT, &mut proc_task).await {
            // On timeout, all is well
            Err(_) => (),
            // The process returned an error
            Ok(Ok(Err(error))) => return Err(error),
            Ok(Ok(Ok(()))) => unreachable!("process monitor stopped prematurely"),
            Ok(Err(_)) => unreachable!("process monitor panicked"),
        }

        Ok(ProcessMonitorHandle {
            stop_proc_tx: Some(stop_proc_tx),
            proc_task,
            states,
        })
    }
}

/// Run until the process exits or `stop_rx` is signaled
async fn handle_eslogger_output(
    mut proc: tokio::process::Child,
    states: ProcessStates,
    stop_rx: oneshot::Receiver<oneshot::Sender<()>>,
) -> Result<(), Error> {
    let stdout = proc.stdout.take().unwrap();
    let stderr = proc.stderr.take().unwrap();

    // Parse each line from stdout as an ESMessage
    tokio::spawn(async move {
        let reader = BufReader::new(stdout);
        let mut lines = reader.lines();

        while let Ok(Some(line)) = lines.next_line().await {
            // Each line from eslogger is a JSON object, one of several types of messages;
            // see `ESMessage`
            let val: ESMessage = match serde_json::from_str(&line) {
                Ok(val) => val,
                Err(error) => {
                    log::error!("Failed to parse eslogger message: {error}");
                    continue;
                }
            };

            let mut inner = states.inner.lock().unwrap();
            inner.handle_message(val);
        }
    });

    // Store the most recent stderr line in case we need to return an error
    let last_stderr = tokio::spawn(async move {
        let reader = BufReader::new(stderr);
        let mut lines = reader.lines();
        let mut last_error = None;

        while let Ok(Some(line)) = lines.next_line().await {
            last_error = Some(line);
        }
        last_error
    });

    // Wait for a stop signal or process exit
    let result = tokio::select! {
        result = proc.wait() => {
            match result {
                Ok(status) => {
                    if let Ok(Some(last_error)) = last_stderr.await {
                        log::error!("eslogger error: {last_error}");
                        if let Some(error) = parse_eslogger_error(&last_error) {
                            return Err(error);
                        }
                    }
                    Err(Error::MonitorFailed(io::Error::other(format!("eslogger stopped unexpectedly: {status}"))))
                }
                Err(error) => Err(Error::MonitorFailed(error)),
            }
        }
        Ok(response_tx) = stop_rx => {
            if let Err(error) = proc.kill().await {
                log::error!("Failed to kill eslogger: {error}");
            }
            if tokio::time::timeout(SHUTDOWN_TIMEOUT, proc.wait())
                .await
                .is_err()
            {
                log::error!("Failed to wait for ST process handler");
            }
            let _ = response_tx.send(());

            Ok(())
        }
    };

    log::debug!("Process monitor stopped");

    result
}

/// Launch a new instance of `eslogger`, listening for exec, fork, and exit syscalls
fn spawn_eslogger() -> Result<tokio::process::Child, Error> {
    let mut cmd = tokio::process::Command::new("/usr/bin/eslogger");
    cmd.args(["exec", "fork", "exit"])
        .kill_on_drop(true)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped());
    cmd.spawn().map_err(Error::StartMonitor)
}

impl ProcessMonitorHandle {
    pub async fn shutdown(&mut self) {
        let Some(stop_tx) = self.stop_proc_tx.take() else {
            return;
        };

        let (tx, rx) = oneshot::channel();
        let _ = stop_tx.send(tx);
        let _ = rx.await;
    }

    pub async fn wait(&mut self) -> Result<(), Error> {
        (&mut self.proc_task)
            .await
            .map_err(Error::MonitorTaskPanicked)?
    }

    pub fn states(&self) -> &ProcessStates {
        &self.states
    }
}

/// Controls the known exclusion states of all processes
#[derive(Debug, Clone)]
pub struct ProcessStates {
    inner: Arc<Mutex<InnerProcessStates>>,
}

/// Possible states of each process
#[derive(Debug, Clone)]
pub enum ExclusionStatus {
    /// The process should be excluded from the VPN
    Excluded,
    /// The process should not be excluded from the VPN
    Included,
    /// The process is unknown
    Unknown,
}

#[derive(Debug)]
struct InnerProcessStates {
    processes: HashMap<u32, ProcessInfo>,
    exclude_paths: HashSet<PathBuf>,
}

impl ProcessStates {
    /// Initialize process states
    fn new() -> Result<Self, Error> {
        let mut states = InnerProcessStates {
            processes: HashMap::new(),
            exclude_paths: HashSet::new(),
        };

        let processes = list_pids().map_err(Error::InitializePids)?;

        for pid in processes {
            let path = process_path(pid).map_err(|error| Error::FindProcessPath(error, pid))?;
            states.processes.insert(pid, ProcessInfo::included(path));
        }

        Ok(ProcessStates {
            inner: Arc::new(Mutex::new(states)),
        })
    }

    pub fn exclude_paths(&self, paths: HashSet<PathBuf>) {
        let mut inner = self.inner.lock().unwrap();

        for info in inner.processes.values_mut() {
            // Remove no-longer excluded paths from exclusion list
            let mut new_exclude_paths: HashSet<_> = info
                .excluded_by_paths
                .intersection(&paths)
                .cloned()
                .collect();

            // Check if own path is excluded
            if paths.contains(&info.exec_path) && !new_exclude_paths.contains(&info.exec_path) {
                new_exclude_paths.insert(info.exec_path.to_owned());
            }

            info.excluded_by_paths = new_exclude_paths;
        }

        inner.exclude_paths = paths;
    }

    pub fn get_process_status(&self, pid: u32) -> ExclusionStatus {
        let inner = self.inner.lock().unwrap();
        match inner.processes.get(&pid) {
            Some(val) if val.is_excluded() => ExclusionStatus::Excluded,
            Some(_) => ExclusionStatus::Included,
            None => ExclusionStatus::Unknown,
        }
    }
}

impl InnerProcessStates {
    fn handle_message(&mut self, msg: ESMessage) {
        let pid = msg.process.audit_token.pid;

        match msg.event {
            ESEvent::Fork(evt) => self.handle_fork(pid, msg.process.executable.path, evt),
            ESEvent::Exec(evt) => self.handle_exec(pid, evt),
            ESEvent::Exit {} => self.handle_exit(pid),
        }
    }

    // For new processes, inherit all exclusion state from the parent, if there is one.
    // Otherwise, look up excluded paths
    fn handle_fork(&mut self, parent_pid: u32, exec_path: PathBuf, msg: ESForkEvent) {
        let pid = msg.child.audit_token.pid;

        if self.processes.contains_key(&pid) {
            log::error!("Conflicting pid! State already contains {pid}");
        }

        // Inherit exclusion status from parent
        let base_info = match self.processes.get(&parent_pid) {
            Some(parent_info) => parent_info.to_owned(),
            None => {
                log::error!("{pid}: Unknown parent pid {parent_pid}!");
                ProcessInfo::included(exec_path)
            }
        };

        // no exec yet; only pid and parent pid change
        if base_info.is_excluded() {
            log::trace!(
                "{pid} excluded (inherited from {parent_pid}) (exclude paths: {:?}",
                base_info.excluded_by_paths
            );
        }

        self.processes.insert(pid, base_info);
    }

    fn handle_exec(&mut self, pid: u32, msg: ESExecEvent) {
        let Some(info) = self.processes.get_mut(&pid) else {
            log::error!("exec received for unknown pid {pid}");
            return;
        };

        info.exec_path = PathBuf::from(msg.dyld_exec_path);

        // If the path is already excluded, no need to add it again
        if info.excluded_by_paths.contains(&info.exec_path) {
            return;
        }

        // Exclude if path is excluded
        if self.exclude_paths.contains(&info.exec_path) {
            info.excluded_by_paths.insert(info.exec_path.to_owned());
            log::trace!("Excluding {pid} by path: {}", info.exec_path.display());
        }
    }

    fn handle_exit(&mut self, pid: u32) {
        if self.processes.remove(&pid).is_none() {
            log::error!("exit syscall for unknown pid {pid}");
        }
    }
}

/// Obtain a list of all pids
fn list_pids() -> io::Result<Vec<u32>> {
    // SAFETY: Passing in null and 0 returns the number of processes
    let num_pids = unsafe { proc_listallpids(ptr::null_mut(), 0) };
    if num_pids <= 0 {
        return Err(io::Error::last_os_error());
    }
    let num_pids = usize::try_from(num_pids).unwrap();
    let mut pids = vec![0u32; num_pids];

    let buf_sz = (num_pids * std::mem::size_of::<u32>()) as i32;
    // SAFETY: 'pids' is large enough to contain 'num_pids' processes
    let num_pids = unsafe { proc_listallpids(pids.as_mut_ptr() as *mut c_void, buf_sz) };
    if num_pids == -1 {
        return Err(io::Error::last_os_error());
    }

    pids.resize(usize::try_from(num_pids).unwrap(), 0);

    Ok(pids)
}

fn process_path(pid: u32) -> io::Result<PathBuf> {
    let mut buffer = [0u8; libc::MAXPATHLEN as usize];
    // SAFETY: `proc_pidpath` returns at most `buffer.len()` bytes
    let buf_len = unsafe {
        proc_pidpath(
            pid as i32,
            buffer.as_mut_ptr() as *mut c_void,
            buffer.len() as u32,
        )
    };
    if buf_len == -1 {
        return Err(io::Error::last_os_error());
    }
    Ok(PathBuf::from(
        std::str::from_utf8(&buffer[0..buf_len as usize])
            .map_err(|_| io::Error::new(io::ErrorKind::InvalidInput, "invalid process path"))?,
    ))
}

#[derive(Debug, Clone)]
struct ProcessInfo {
    exec_path: PathBuf,
    excluded_by_paths: HashSet<PathBuf>,
}

impl ProcessInfo {
    fn included(exec_path: PathBuf) -> Self {
        ProcessInfo {
            exec_path,
            excluded_by_paths: HashSet::new(),
        }
    }

    fn is_excluded(&self) -> bool {
        !self.excluded_by_paths.is_empty()
    }
}

/// `fork` event details
#[derive(Debug, Deserialize)]
struct ESForkChild {
    audit_token: ESAuditToken,
}

/// `fork` event returned by `eslogger`
#[derive(Debug, Deserialize)]
struct ESForkEvent {
    child: ESForkChild,
}

/// `exec` event returned by `eslogger`
#[derive(Debug, Deserialize)]
struct ESExecEvent {
    dyld_exec_path: String,
}

/// Event that triggered the message returned by `eslogger`.
/// See the `es_events_t` struct for more information:
/// https://developer.apple.com/documentation/endpointsecurity/es_message_t/3228969-event?language=objc
/// A list of all event types can be found here:
/// https://developer.apple.com/documentation/endpointsecurity/es_event_type_t/es_event_type_notify_fork?language=objc
#[derive(Debug, Deserialize)]
#[serde(rename_all = "lowercase")]
enum ESEvent {
    Fork(ESForkEvent),
    Exec(ESExecEvent),
    Exit {},
}

/// Message containing the path to the image of the process.
/// This message is analogous to the `executable` field of `es_process_t`:
/// https://developer.apple.com/documentation/endpointsecurity/es_process_t/3228975-audit_token?language=objc
#[derive(Debug, Deserialize)]
struct ESExecutable {
    path: PathBuf,
}

/// Message containing the process identifier of the process.
/// This message is analogous to the `audit_token` field of `es_process_t`:
/// https://developer.apple.com/documentation/endpointsecurity/es_process_t/3228975-audit_token?language=objc
#[derive(Debug, Deserialize)]
struct ESAuditToken {
    pid: u32,
}

/// Process information for the message returned by `eslogger`.
/// This message is analogous to the `es_process_t` struct:
/// https://developer.apple.com/documentation/endpointsecurity/es_process_t?language=objc
#[derive(Debug, Deserialize)]
struct ESProcess {
    audit_token: ESAuditToken,
    executable: ESExecutable,
}

/// This struct represents each message returned by eslogger
/// This message is analogous to the `es_message_t` struct:
/// https://developer.apple.com/documentation/endpointsecurity/es_message_t?language=objc
#[derive(Debug, Deserialize)]
struct ESMessage {
    event: ESEvent,
    process: ESProcess,
}

fn parse_eslogger_error(stderr_str: &str) -> Option<Error> {
    if stderr_str.contains("ES_NEW_CLIENT_RESULT_ERR_NOT_PERMITTED") {
        Some(Error::NeedFullDiskPermissions)
    } else {
        None
    }
}

/// Check whether the current macOS version is supported, and return an error otherwise
fn check_os_version_support() -> Result<(), Error> {
    match MacosVersion::new() {
        Ok(version) => check_os_version_support_inner(version),
        Err(error) => {
            log::error!("Failed to detect macOS version: {error}");
            Ok(())
        }
    }
}

fn check_os_version_support_inner(version: MacosVersion) -> Result<(), Error> {
    if version >= *MIN_OS_VERSION {
        Ok(())
    } else {
        Err(Error::UnsupportedMacosVersion { actual: version })
    }
}

#[test]
fn test_min_os_version() {
    assert!(check_os_version_support_inner(MIN_OS_VERSION.clone()).is_ok());

    // test unsupported version
    assert!(
        check_os_version_support_inner(MacosVersion::from_raw_version("12.1").unwrap()).is_err()
    );

    // test supported version
    assert!(
        check_os_version_support_inner(MacosVersion::from_raw_version("13.0").unwrap()).is_ok()
    );
}
