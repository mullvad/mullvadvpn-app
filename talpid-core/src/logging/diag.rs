//! Additional logging that may be useful

/// Additional logging for Windows
#[cfg(target_os = "windows")]
pub mod windows {
    use anyhow::Context;
    use anyhow::bail;
    use std::ffi::CStr;
    use std::ffi::OsString;
    use std::fmt::Write;
    use std::os::windows::ffi::OsStringExt;
    use std::path::Path;
    use talpid_windows::env::get_system_dir;
    use talpid_windows::string::multibyte_to_wide;
    use tokio::io::AsyncWriteExt;
    use tokio::process::Command;
    use windows_sys::Win32::Globalization::CP_ACP;

    /// Keywords used to filter output
    const KEYWORDS: &[&str] = &[
        "wireguard",
        "wintun",
        "tunnel",
        "mullvad",
        "split-tunnel",
        "split tunnel",
    ];

    /// Dump logs about tunnel devices and relevant drivers
    ///
    /// Currently, this will log the output of `pnputil` and `driverquery`, with filtering.
    pub async fn log_device_info(log_dir: &Path) -> anyhow::Result<()> {
        use crate::logging::rotate_log;
        use tokio::{fs::File, io::BufWriter};

        const TIMESTAMP_FMT: &str = "%Y-%m-%d %H:%M:%S";

        let log_path = log_dir.join("device.log");

        // why u gotta be 'static
        let log_path2 = log_path.clone();
        tokio::task::spawn_blocking(move || rotate_log(&log_path2))
            .await
            .unwrap()
            .context("Failed to rotate log")?;

        let logger = File::options()
            .write(true)
            .create(true)
            .truncate(true)
            .open(log_path)
            .await
            .context("Failed to open device log")?;

        let mut logger = BufWriter::new(logger);

        // Log the current time
        logger
            .write_all(
                format!(
                    "Log time: {}\n\n",
                    chrono::Local::now().format(TIMESTAMP_FMT)
                )
                .as_bytes(),
            )
            .await?;

        async fn run_cmd_and_write_logs(
            logger: &mut BufWriter<File>,
            cmd: &mut Command,
            parse_output: impl FnOnce(String) -> anyhow::Result<String>,
        ) -> anyhow::Result<()> {
            let logs = run_cmd(cmd).await.and_then(parse_output);
            logger.write_all(format_logs(cmd, logs)?.as_bytes()).await?;
            Ok(())
        }

        run_cmd_and_write_logs(&mut logger, &mut driverquery_cmd()?, parse_driverquery).await?;
        run_cmd_and_write_logs(&mut logger, &mut pnputil_cmd()?, parse_pnputil).await?;
        run_cmd_and_write_logs(
            &mut logger,
            &mut pnputil_problem_cmd()?,
            parse_pnputil_problem,
        )
        .await?;

        let _ = logger.flush().await;

        Ok(())
    }

    /// Run `cmd` and collect its output
    async fn run_cmd(cmd: &mut Command) -> anyhow::Result<String> {
        let out = cmd.output().await.context("Failed to run driverquery")?;

        if !out.status.success() {
            bail!("driverquery failed: {:?}", out.status.code());
        }

        parse_raw_cmd_output(&out.stdout)
    }

    fn format_logs(cmd: &Command, output: anyhow::Result<String>) -> anyhow::Result<String> {
        let mut buf = String::new();

        writeln!(&mut buf, "{} (filtered)", format_command(cmd)?)?;
        writeln!(&mut buf, "--------")?;

        match output {
            Ok(out) => {
                writeln!(&mut buf, "{out}")?;
            }
            Err(err) => {
                writeln!(&mut buf, "The command failed due to an error: {err}")?;
            }
        }
        writeln!(&mut buf, "--------")?;

        Ok(buf)
    }

    fn parse_driverquery(out_s: String) -> anyhow::Result<String> {
        const KEEP_FIELDS: &[&str] = &[
            "\"Module Name\"",
            "\"Display Name\"",
            "\"Description\"",
            "\"Driver Type\"",
            "\"Start Mode\"",
            "\"Description\"",
            "\"State\"",
            "\"Status\"",
            "\"Path\"",
        ];
        parse_csv(&out_s, KEEP_FIELDS, keyword_filter_fn)
    }

    fn parse_pnputil(out_s: String) -> anyhow::Result<String> {
        const KEEP_FIELDS: &[&str] = &[
            "InstanceId",
            "DeviceDescription",
            "Status",
            "ProblemCode",
            "ProblemStatus",
            "DriverName",
        ];
        parse_csv(&out_s, KEEP_FIELDS, keyword_filter_fn)
    }

    fn parse_pnputil_problem(out_s: String) -> anyhow::Result<String> {
        const KEEP_FIELDS: &[&str] = &[
            "InstanceId",
            "DeviceDescription",
            "Status",
            "ProblemCode",
            "ProblemStatus",
            "DriverName",
        ];
        // In this case, we keep all entries
        parse_csv(&out_s, KEEP_FIELDS, |_| true)
    }

    fn keyword_filter_fn(line: &str) -> bool {
        KEYWORDS.iter().any(|word| {
            line.to_ascii_lowercase()
                .contains(&word.to_ascii_lowercase())
        })
    }

    fn parse_raw_cmd_output(bytes: &[u8]) -> anyhow::Result<String> {
        // Convert from current codepage to UTF8
        // Seems not entirely correct, but probably good enough
        let mut bytes = bytes.to_vec();
        bytes.push(0);
        let bytes_cstr = CStr::from_bytes_until_nul(&bytes).unwrap();

        let str = multibyte_to_wide(bytes_cstr, CP_ACP).context("Invalid pnputil output")?;
        let out_s = OsString::from_wide(&str);
        Ok(out_s.to_string_lossy().into_owned())
    }

    fn parse_csv(
        s: &str,
        keep_fields: &[&str],
        entry_filter_fn: impl Fn(&str) -> bool,
    ) -> anyhow::Result<String> {
        let mut lines = s.lines();
        // The first line contains the columns, so always keep it
        let first_line = lines.next().context("Empty output")?;

        let fields_to_keep: Vec<_> = first_line
            .split(',')
            .enumerate()
            .filter_map(|(ind, field)| {
                if keep_fields
                    .iter()
                    .any(|keep_f| field.eq_ignore_ascii_case(keep_f))
                {
                    return Some(ind);
                }
                None
            })
            .collect();

        let filtered_lines: Vec<_> = std::iter::once(first_line)
            // Filter out entries using entry_filter_fn
            .chain(lines.filter(|line| entry_filter_fn(line)))
            // Keep only fields of interest
            .map(|line| {
                line.split(',').enumerate().filter_map(|(ind, field)| {
                    if fields_to_keep.contains(&ind) {
                        return Some(field);
                    }
                    None
                }).collect::<Vec<_>>().join(",")
            })
            .collect();

        Ok(filtered_lines.join("\n"))
    }

    fn driverquery_cmd() -> anyhow::Result<Command> {
        let path = get_system_dir()?.join("driverquery.exe");
        let mut driver_query = Command::new(path);
        driver_query.args(["/FO", "csv", "/V"]);
        Ok(driver_query)
    }

    fn pnputil_cmd() -> anyhow::Result<Command> {
        let path = get_system_dir()?.join("pnputil.exe");
        let mut pnputil = Command::new(path);
        // Enumerate network devices
        pnputil.args([
            "/enum-devices",
            "/class",
            "{4d36e972-e325-11ce-bfc1-08002be10318}",
        ]);
        pnputil.args(["/format", "csv"]);
        Ok(pnputil)
    }

    fn pnputil_problem_cmd() -> anyhow::Result<Command> {
        let path = get_system_dir()?.join("pnputil.exe");
        let mut pnputil = Command::new(path);
        // Enumerate devices with issues
        pnputil.args(["/enum-devices", "/problem"]);
        pnputil.args(["/format", "csv"]);
        Ok(pnputil)
    }

    fn format_command(cmd: &Command) -> anyhow::Result<String> {
        let mut s = String::new();

        let prog = Path::new(cmd.as_std().get_program())
            .file_name()
            .context("Missing command filename")?
            .display();
        write!(&mut s, r#"{prog}"#)?;

        for arg in cmd.as_std().get_args() {
            write!(&mut s, r#" "{}""#, arg.display())?;
        }

        Ok(s)
    }

    #[cfg(test)]
    mod test {
        use super::*;

        /// Test whether driverquery output is filtered correctly
        #[tokio::test]
        async fn test_driverquery_output() {
            let test_output_path = Path::new(&std::env::var("CARGO_MANIFEST_DIR").unwrap())
                .join("src/logging/driverquery-out.testdata");

            // Uncomment to generate new output
            //tokio::fs::write(
            //    &test_output_path,
            //    driverquery_cmd().unwrap().output().await.unwrap().stdout,
            //)
            //.await
            //.unwrap();

            let my_output = std::fs::read(test_output_path).unwrap();
            let my_output = parse_raw_cmd_output(&my_output).unwrap();
            let parsed_driverquery_output = parse_driverquery(my_output).unwrap();
            let formatted_driverquery_log =
                format_logs(&driverquery_cmd().unwrap(), Ok(parsed_driverquery_output)).unwrap();

            insta::assert_snapshot!(formatted_driverquery_log);
        }

        /// Test whether pnputil output is filtered correctly
        #[tokio::test]
        async fn test_pnputil_output() {
            let test_output_path = Path::new(&std::env::var("CARGO_MANIFEST_DIR").unwrap())
                .join("src/logging/pnputil-out.testdata");

            // Uncomment to generate new output
            //tokio::fs::write(
            //    &test_output_path,
            //    pnputil_cmd().unwrap().output().await.unwrap().stdout,
            //)
            //.await
            //.unwrap();

            let my_output = std::fs::read(test_output_path).unwrap();
            let my_output = parse_raw_cmd_output(&my_output).unwrap();
            let parsed_output = parse_pnputil(my_output).unwrap();
            let formatted_logs = format_logs(&pnputil_cmd().unwrap(), Ok(parsed_output)).unwrap();

            insta::assert_snapshot!(formatted_logs);
        }

        /// Test whether pnputil output is filtered correctly
        #[tokio::test]
        async fn test_pnputil_problem_output() {
            let test_output_path = Path::new(&std::env::var("CARGO_MANIFEST_DIR").unwrap())
                .join("src/logging/pnputil-problem-out.testdata");

            // Uncomment to generate new output
            //tokio::fs::write(
            //    &test_output_path,
            //    pnputil_problem_cmd().unwrap().output().await.unwrap().stdout,
            //)
            //.await
            //.unwrap();

            let my_output = std::fs::read(test_output_path).unwrap();
            let my_output = parse_raw_cmd_output(&my_output).unwrap();
            let parsed_output = parse_pnputil_problem(my_output).unwrap();
            let formatted_logs =
                format_logs(&pnputil_problem_cmd().unwrap(), Ok(parsed_output)).unwrap();

            insta::assert_snapshot!(formatted_logs);
        }
    }
}
