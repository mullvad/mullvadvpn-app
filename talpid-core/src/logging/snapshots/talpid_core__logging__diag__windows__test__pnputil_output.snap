---
source: talpid-core/src/logging/diag.rs
expression: formatted_logs
snapshot_kind: text
---
pnputil.exe "/enum-devices" "/class" "{4d36e972-e325-11ce-bfc1-08002be10318}" "/format" "csv" (filtered)
--------
InstanceId,DeviceDescription,Status,ProblemCode,ProblemStatus,DriverName
ROOT\WINTUN\0000,Wintun Userspace Tunnel,Started,,,oem2.inf

--------
